File: integration-sample/cpp/central_logs.cpp
// central-logs integration client — batched NDJSON inserts (C++17).
//
// Dependencies: libcurl, pthreads.
//   Build:  g++ -std=c++17 app.cpp central_logs.cpp -o app -lcurl -lpthread
//   Demo:   g++ -std=c++17 -DCL_DEMO central_logs.cpp -o cl-demo -lcurl -lpthread
//
// Setup (see INTEGRATION.md):
//   1. Admin key pinned in the server's .env (CENTRAL_LOGS_HTTP_API_KEY).
//   2. Mint an insert-only key (POST /v1/api-keys, scopes: "insert").
//   3. Export for this app:
//        CENTRAL_LOGS_URL      (default http://localhost:8080)
//        CENTRAL_LOGS_API_KEY  (the raw clk_... value — required)
//        CENTRAL_LOGS_SERVICE  (default "default-app")
//
// Usage:
//   centrallogs::init();
//   centrallogs::log(centrallogs::Level::Info, "app started", {
//       {"version", "1.2.3"},
//       {"pid",     centrallogs::raw(std::to_string(getpid()))},
//   });
//   centrallogs::shutdown();  // final flush
//
// A background std::thread flushes every CENTRAL_LOGS_FLUSH_MS (default
// 1000). On persistent failure the batch goes to stderr and the queue is
// bounded, so outages can't exhaust memory.

#include "central_logs.hpp"

#include <curl/curl.h>

#include <atomic>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <deque>
#include <mutex>
#include <thread>
#include <vector>

namespace centrallogs {

namespace {

std::string g_url, g_key, g_service;
long g_flush_ms = 1000;

std::mutex g_mutex;
std::deque<std::string> g_queue;
std::atomic<bool> g_closed{false};
std::atomic<long> g_dropped{0};
std::thread g_thread;

constexpr size_t kMaxQueue = 5000;
constexpr size_t kMaxBatch = 100;

std::string env_or(const char *key, const char *def) {
    const char *v = std::getenv(key);
    return (v && *v) ? std::string(v) : std::string(def);
}

std::string timestamp_now() {
    using namespace std::chrono;
    const auto now = system_clock::now();
    const auto secs = time_point_cast<seconds>(now);
    const auto ms = duration_cast<milliseconds>(now - secs).count();
    std::time_t t = system_clock::to_time_t(now);
    std::tm tm{};
    gmtime_r(&t, &tm);
    char buf[40];
    std::snprintf(buf, sizeof buf, "%04d-%02d-%02dT%02d:%02d:%02d.%03lldZ",
                  tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
                  tm.tm_hour, tm.tm_min, tm.tm_sec, static_cast<long long>(ms));
    return buf;
}

const char *level_name(Level l) {
    switch (l) {
    case Level::Debug: return "debug";
    case Level::Warn:  return "warn";
    case Level::Error: return "error";
    case Level::Fatal: return "fatal";
    default:           return "info";
    }
}

} // namespace

std::string json_escape(const std::string &s) {
    std::string out;
    out.reserve(s.size() + 2);
    out += '"';
    for (unsigned char c : s) {
        switch (c) {
        case '"':  out += "\\\""; break;
        case '\\': out += "\\\\"; break;
        case '\n': out += "\\n";  break;
        case '\r': out += "\\r";  break;
        case '\t': out += "\\t";  break;
        default:
            if (c < 0x20) {
                char buf[8];
                std::snprintf(buf, sizeof buf, "\\u%04x", c);
                out += buf;
            } else {
                out += static_cast<char>(c);
            }
        }
    }
    out += '"';
    return out;
}

void init(const std::string &base_url) {
    g_url = base_url.empty() ? env_or("CENTRAL_LOGS_URL", "http://localhost:8080") : base_url;
    while (!g_url.empty() && g_url.back() == '/') g_url.pop_back();
    g_key = env_or("CENTRAL_LOGS_API_KEY", "");
    g_service = env_or("CENTRAL_LOGS_SERVICE", "default-app");
    g_flush_ms = std::atol(env_or("CENTRAL_LOGS_FLUSH_MS", "1000").c_str());
    if (g_flush_ms < 100) g_flush_ms = 100;

    g_closed = false;
    g_thread = std::thread([] {
        while (!g_closed.load()) {
            std::this_thread::sleep_for(std::chrono::milliseconds(g_flush_ms));
            if (g_closed.load()) break;
            flush();
        }
    });
}

void log(Level level, const std::string &msg) { log(level, msg, {}); }

void log(Level level, const std::string &msg,
         const std::map<std::string, std::string> &fields) {
    if (g_closed.load()) return;

    std::string line = "{\"service\":" + json_escape(g_service) +
                       ",\"level\":\"" + level_name(level) +
                       "\",\"msg\":" + json_escape(msg) +
                       ",\"ts\":\"" + timestamp_now() + "\"";
    for (const auto &kv : fields) {
        // A value wrapped in Raw{} is emitted verbatim (see raw() helper —
        // detected by the reserved prefix below).
        static const std::string kRawPrefix = "\x01RAW\x01";
        if (kv.second.rfind(kRawPrefix, 0) == 0) {
            line += ",\"" + kv.first + "\":" + kv.second.substr(kRawPrefix.size());
        } else {
            line += ",\"" + kv.first + "\":" + json_escape(kv.second);
        }
    }
    line += "}";

    size_t size;
    {
        std::lock_guard<std::mutex> lk(g_mutex);
        if (g_queue.size() >= kMaxQueue) g_queue.pop_front(); // bounded
        g_queue.push_back(std::move(line));
        size = g_queue.size();
    }
    if (size >= kMaxBatch) flush();
}

size_t queue_size() {
    std::lock_guard<std::mutex> lk(g_mutex);
    return g_queue.size();
}

namespace {
size_t curl_write(char *ptr, size_t size, size_t nmemb, void *ud) {
    (void)ptr; (void)ud;
    return size * nmemb;
}
} // namespace

void flush() {
    std::deque<std::string> batch;
    {
        std::lock_guard<std::mutex> lk(g_mutex);
        if (g_queue.empty()) return;
        batch.swap(g_queue);
    }

    if (g_key.empty()) {
        g_dropped += batch.size();
        std::fprintf(stderr, "[central-logs] flush failed (%ld dropped total): CENTRAL_LOGS_API_KEY not set\n",
                     g_dropped.load());
        for (const auto &line : batch) std::fprintf(stderr, "%s\n", line.c_str());
        return;
    }

    std::string body;
    body.reserve(batch.size() * 128);
    for (const auto &line : batch) {
        body += line;
        body += '\n';
    }

    CURL *curl = curl_easy_init();
    if (!curl) {
        g_dropped += batch.size();
        std::fprintf(stderr, "[central-logs] curl init failed\n");
        return;
    }
    const std::string url = g_url + "/v1/logs";
    const std::string auth = "Authorization: Bearer " + g_key;
    curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/x-ndjson");
    headers = curl_slist_append(headers, auth.c_str());
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, curl_write);

    CURLcode rc = curl_easy_perform(curl);
    long status = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);

    if (rc != CURLE_OK || status != 200) {
        g_dropped += batch.size();
        std::fprintf(stderr, "[central-logs] flush failed (%ld dropped total): %sHTTP %ld\n",
                     g_dropped.load(), rc == CURLE_OK ? "" : curl_easy_strerror(rc), status);
        for (const auto &line : batch) std::fprintf(stderr, "%s\n", line.c_str());
    }
}

void shutdown() {
    if (g_closed.exchange(true)) return;
    if (g_thread.joinable()) g_thread.join();
    flush();
}

} // namespace centrallogs

#ifdef CL_DEMO
int main() {
    centrallogs::init();
    centrallogs::log(centrallogs::Level::Info, "central-logs c++ client smoke test");
    centrallogs::log(centrallogs::Level::Warn, "queue lag 12s", {
        {"duration_ms", centrallogs::raw("12000")},
        {"queue", "events"},
    });
    centrallogs::shutdown();
    std::puts("flushed; verify with: curl '.../api/logs?filter=service:default-app&window=5m'");
    return 0;
}
#endif

File: integration-sample/js/central-logs.ts
/**
 * central-logs integration — typed logger that batches NDJSON into
 * `POST {BASE_URL}/v1/logs` and survives transient ingest failures.
 *
 * ─── Setup (one per app) ─────────────────────────────────────────────────
 *   1. Set an admin key on the server, pinned in the server's `.env`:
 *
 *        CENTRAL_LOGS_HTTP_API_KEY=clk_$(openssl rand -hex 24)
 *
 *      and restart the server. This is the long-lived admin credential.
 *
 *   2. Mint an insert-only key for THIS app:
 *
 *        curl -sf -X POST http://localhost:8080/v1/api-keys \
 *          -H "Authorization: Bearer $ADMIN_KEY" \
 *          -H 'Content-Type: application/json' \
 *          -d '{"name":"my-app","scopes":"insert"}'
 *
 *      The raw `clk_...` value is shown exactly once. Persist it next to
 *      your app's other secrets — `.env`, secrets manager, etc.
 *
 *   3. Set these env vars for this app:
 *
 *        CENTRAL_LOGS_URL=http://localhost:8080        # default
 *        CENTRAL_LOGS_API_KEY=clk_...                  # from step 2
 *        CENTRAL_LOGS_SERVICE=my-app                   # filter key, default "default-apps"
 *
 * ─── Usage ────────────────────────────────────────────────────────────────
 *
 *      import { log, logApi, withCentralLogging, flush } from "./central-logs.js";
 *
 *      log("info", "service started", { version: "1.2.3" });
 *      log("warn", "queue lag 12s",   { queue: "events", duration_ms: 12000 });
 *
 *      // Bun/Hono/Express-style wrapped route:
 *      const routes = withCentralLogging({
 *        "/api/health": async () => new Response("ok"),
 *        "/api/users":  async (req) => fetch(...),
 *      });
 *
 *      // Force a final flush on shutdown:
 *      process.on("SIGTERM", () => { void flush(); });
 *
 * Records are batched every FLUSH_MS (default 1000ms) or once MAX_BATCH
 * lines accumulate, whichever comes first. If the server is unreachable the
 * batch is logged to stderr and dropped (the queue is bounded by MAX_QUEUE,
 * default 5000, to avoid unbounded memory growth in a long outage).
 *
 * ─── Field schema sent to central-logs ───────────────────────────────────
 *   service      string (required) — CENTRAL_LOGS_SERVICE
 *   level        "info" | "warn" | "error"
 *   msg          string — short free-text summary
 *   ts           RFC3339 string — set automatically
 *   duration_ms  number  — auto-added by withCentralLogging
 *   status       number  — auto-added by withCentralLogging
 *   ...          anything you pass in `fields` lands in the JSON `attributes`
 *                column and becomes queryable via the filter DSL.
 *                Common typed fields (promote via --hot-attribute for
 *                zonemap-pruned scans): user_id:bigint, trace_id:varchar,
 *                request_route:varchar, status:bigint.
 */

// ── config ─────────────────────────────────────────────────────────────────
const BASE_URL = (process.env.CENTRAL_LOGS_URL ?? "http://localhost:8080").replace(/\/+$/, "");
const API_KEY = process.env.CENTRAL_LOGS_API_KEY ?? "";
const SERVICE = process.env.CENTRAL_LOGS_SERVICE ?? "default-apps";
const FLUSH_MS = Number(process.env.CENTRAL_LOGS_FLUSH_MS ?? 1000);
const MAX_BATCH = Number(process.env.CENTRAL_LOGS_MAX_BATCH ?? 100);
const MAX_QUEUE = Number(process.env.CENTRAL_LOGS_MAX_QUEUE ?? 5000);

export type Level = "info" | "warn" | "error";

export interface LogRecord {
  service: string;
  level: Level;
  msg: string;
  ts: string;
  [key: string]: unknown;
}

// ── queue + flush ──────────────────────────────────────────────────────────
// Config is re-read on every flush so callers can mutate process.env after
// importing this module (e.g. dotenv-loaded apps, test harnesses).
function readApiKey(): string {
  return process.env.CENTRAL_LOGS_API_KEY ?? "";
}

const queue: LogRecord[] = [];
let flushing = false;
let dropped = 0;

function enqueue(record: LogRecord) {
  if (!readApiKey()) {
    // Misconfiguration is loud and visible — fall back to stderr so
    // developers notice, but don't crash the app.
    console.error(JSON.stringify(record));
    return;
  }
  if (queue.length >= MAX_QUEUE) queue.shift(); // bounded: drop oldest
  queue.push(record);
  if (queue.length >= MAX_BATCH) void flush();
}

export function log(level: Level, msg: string, fields: Record<string, unknown> = {}) {
  enqueue({ service: SERVICE, level, msg, ts: new Date().toISOString(), ...fields });
}

/** Convenience for logging a single Request/Response pair (Bun, Hono, Workers).
 *  Auto-derives level from status (<400 = info, <500 = warn, else error). */
export function logApi(req: Request, fields: Record<string, unknown> = {}) {
  const status = typeof fields.status === "number" ? fields.status : 0;
  const level: Level = status >= 500 ? "error" : status >= 400 ? "warn" : "info";
  log(level, `${req.method} ${new URL(req.url).pathname}`, {
    status: status || undefined,
    ...fields,
  });
}

/** Wrap a route map to log every call with timing + status + query string. */
export function withCentralLogging(routes: Record<string, (req: Request) => Promise<Response>>) {
  for (const [path, handler] of Object.entries(routes)) {
    if (typeof handler !== "function") continue;
    routes[path] = async (req: Request) => {
      const started = Date.now();
      const query = new URL(req.url).search || undefined;
      try {
        const res = await handler(req);
        log(res.status < 400 ? "info" : "warn", `${req.method} ${path}`, {
          status: res.status,
          duration_ms: Date.now() - started,
          query,
        });
        return res;
      } catch (e: unknown) {
        const message = e instanceof Error ? e.message : String(e);
        log("error", `${req.method} ${path} failed: ${message}`, {
          status: 500,
          duration_ms: Date.now() - started,
          query,
        });
        throw e;
      }
    };
  }
  return routes;
}

/** Flush the queue. Safe to await on shutdown. */
export async function flush(): Promise<void> {
  if (flushing || queue.length === 0) return;
  const key = readApiKey();
  if (!key) {
    // Misconfigured at flush time — drain to stderr so the records aren't
    // silently held in memory.
    for (const r of queue) console.error(JSON.stringify(r));
    queue.splice(0, queue.length);
    return;
  }
  flushing = true;
  const batch = queue.splice(0, queue.length);
  const body = batch.map((r) => JSON.stringify(r)).join("\n");
  try {
    const res = await fetch(`${BASE_URL}/v1/logs`, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-ndjson",
        Authorization: `Bearer ${key}`,
      },
      body,
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${(await res.text()).slice(0, 200)}`);
    const out = (await res.json().catch(() => ({}))) as { rejected?: number };
    if (typeof out.rejected === "number" && out.rejected > 0) {
      console.error(`[central-logs] rejected ${out.rejected}/${batch.length} records`);
    }
  } catch (e: unknown) {
    // Bounded backpressure: drop the batch on persistent failure rather
    // than letting memory grow unbounded. The original lines still go to
    // stderr so nothing is silently lost.
    dropped += batch.length;
    const message = e instanceof Error ? e.message : String(e);
    console.error(`[central-logs] flush failed (${dropped} dropped total): ${message}`);
    for (const r of batch) console.error(JSON.stringify(r));
  } finally {
    flushing = false;
  }
}

// ── plumbing ───────────────────────────────────────────────────────────────
type Timer = { unref?: () => void; [Symbol.dispose]?: () => void };
const flushTimer = setInterval(() => void flush(), FLUSH_MS) as unknown as Timer;
flushTimer.unref?.(); // don't keep the event loop alive solely for flushing
for (const sig of ["SIGINT", "SIGTERM"] as const) {
  process.once(sig, () => {
    clearInterval(flushTimer as unknown as number);
    void flush();
  });
}
